client/src/components/SolutionViewer/SolutionViewer.tsx

```diff
@@ -9,7 +9,7 @@ type Props = {
 };
 type Solution = Pick<ChallengeFile, 'ext' | 'contents' | 'fileKey'>;
 
-function SolutionViewer({ challengeFiles, solution }: Props) {
+function SolutionViewer({ challengeFiles, solution }: Props): JSX.Element {
```

Comment:
Sorry to say it after you are done, but can we disable the rule that force to explicit the return type of the function.

If we aren't interacting with contributors that don't know if we add this to please eslint, or the type of the return function is actually import, then think that the explicit type is indicating an issue with their code, may result in weird code to work around the function.

Reply:
A good example is this file 

https://github.com/freeCodeCamp/freeCodeCamp/blob/53164e71f160ac156de2706b77c28ae18e9ca027/client/src/components/formHelpers/form.tsx#L108-L116

Although it looks fine, in the place where `DynamicForm` is used, `hidebutton` is always `false`, and `buttonText` is always true.

So people are using if statements to work around the type checkers, while we don't need to

Reply:
For the most part, we should take advantage of the fact that we use TypeScript; we should use the strictest rules we can. The explicit returns for functions is a very useful rule.

I would suggest we as reviewers/maintainers be proactive with correcting/guiding Campers who submit code with "weird" code. For example, I have a red flag go off in my brain when I see any type assertion, and take time to see why that assertion is there, and if it is even necessary.

Reply:
if this is Rust, I will be all over this and in favor of this, but in TypeScript it has "feature", which makes the return type of function overwrite the infered typed, and not only that if you create a function that except `foo`, and add this code:

```ts
const something = (): randomType => { return foo as const}
```

Can you expect the type of returned value of `something`? it's `randomType` although you explicitly requested it return foo type in the return function, this is way we shouldn't use "feature", that to be honest broken, and use the normal ways to return the function values when needed.

Reply:
TD;LR if TypeScript doesn't have "feature" that overwrite the inferred return type or the explicit type, we should allow this.

Reply:
<img width="246" alt="image" src="https://user-images.githubusercontent.com/63889819/218272254-d085fed7-3b99-4717-b265-f8a5b29bb1c9.png">

That code would throw a compiler error, because your returned value doesn't match the function signature. And that's exactly why explicit return types are important - it ensures that what we are returning is actually what we want to return, rather than a potential developer error.

Reply:
As Naomi said, the whole point of the explicit return is developers **have** to know what they want to return from a function. So, anything that _"accidentally"_ does not return the declared type should throw a type error:

<details>
  <summary>Example</summary>

```typescript
// I know I want a function that always returns a string
function getMatch(): string {
    const str = "a";
    const match = str.match(/b/)?.[0];
    return match;
}
```

Yields:

```bash
Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.
```

</details>

> if this is Rust, I will be all over this and in favor of this

Me too 😁

Reply:
> And that's exactly why explicit return types are important

And that's exactly why explicit return types are **dangerous**, because when someone consumes the functional the message they receive is viewed as a source of truth, but it isn't the truth because TypeScript can and do lie, which is **dangerous** especially in a codebase that half of it isn't TypeScript, and the many types in the other half are typed wrong.

We are asking maintainers and contributors to always return something correct, when they produce a function, but this isn't realistic. 

This is why when we as maintainers know that the function produced is important, we should explicitly put the return of the produced function `as const`, so when it's consumed, the types of the consumer should align with the `as const`, otherwise we get type error in consumption, and the type error will tell directly which part is the wrong one, the production or the consumption.

Reply:
You know what, let's decide a middle ground. this eslint rule can be applied in the API/Backend freely, but it shouldn't be applied in the client side.

If this is accepted, I am happy.

Reply:
We can discuss this in the next dev meeting.

> We are asking maintainers and contributors to always return something correct, when they produce a function, but this isn't realistic.

This is the whole point of TypeScript, though. Say what you expect in, what should come out, and complain when those conditions are not met.

Reply:
> And that's exactly why explicit return types are dangerous, because when someone consumes the function, the message they receive from TypeScript is viewed as a source of truth, but it isn't the truth because TypeScript can and do lie, which is dangerous especially in a codebase that half of it isn't TypeScript, and the many types in the other half are typed wrong.

You can 'trick' TypeScript via double type assertions like

```ts
const fakeNumber = 'pretending to be a number';
const notReallyANumber: number = fakeNumber as unknown as number;
```

but this isn't what this rule is concerned with. It requires us to set the return type and that's perfectly safe.

```ts
function okay(): string | number {
  return 'just a string'
}

function notOkay(): string {
  let ambiguous: string | number = Math.random() < 0.5 ? '': 0;
  return ambiguous
}
```
This a little contrived, but the `okay` function is fine because it's always safe to make a type _less_ specific. You can't get runtime errors this way, since you have to check if the returned data is a string or a number before using it.

That's not true for `notOkay`, since it really could be a string or a number. You are not allowed to narrow the return type - it's a compile-time error.

In both cases TS is giving us type safety, so the question is just "is this helpful?" not "is this dangerous?".

I can attest to it being a bit of a chore, but it does encourage the creation of more readable types. Also, if you're the one creating or updating the functions it should be less of a chore since you (should) understand the change.

Reply:
In a codebase as large as ours, I'm a firm believer that explicit return types help a *lot* more than they hinder.

Reply:
I know that it's hard for me to convey what I mean, and this is prolonged than needed, so you can ignore this to keep things going, but:

- I don't appreciate the concept of repeating the same example, as I don't know it. I do know what you mean, and I think it is not only not worth it, but it's dangerous too, because we with the "big code base" and not bloated one, that define same values multiple times and the same component multiple times, barely know if the types are correct.

```ts
const fakeNumber = 'pretending to be a number';
const notReallyANumber: number = fakeNumber as unknown as number;
```

- This example is flawed in itself, because I didn't bring "tricking" TypeScript, but I brought that TypeScript can lie to use about the return result, and will lie to us.

---

And as I said, if you guys want to apply it, apply it in API and leave the /client away from it. 

If you don't want to do that, just don't do it, but don't bring the concept of "tricking" TypeScript, because this isn't about it, and don't mention that it's helpful for "big" codebase, because it really isn't, the "big" codebase doesn't need another rail guard, that can break and does lie.

Reply:
It's fine to discuss this here.

> TypeScript can lie to us about the return result, and will lie to us.

Rather than talking in the abstract, we should be concrete. Could you point to some code that has the problem you're concerned about? To be specific: cases where you have a function

```ts
function name() {
// ...
}
```
and adding an explicit type like

```ts
function name(): SomeType {
// ...
}
```
causes a problem.

Reply:
If we really want to discuss it, I have opened 2nd PR with buggy code, showing why it's great to use explicit types but it shouldn't be the default 

https://github.com/freeCodeCamp/freeCodeCamp/pull/49368